builtin/providers/aws/data_source_aws_s3_object.go

```diff
@@ -0,0 +1,221 @@
+				Computed: true,
+			},
+			"metadata": &schema.Schema{
+				Type:     schema.TypeMap,
+				Computed: true,
+			},
+			"range": &schema.Schema{
+				Type:     schema.TypeString,
+				Optional: true,
+			},
+			"server_side_encryption": &schema.Schema{
+				Type:     schema.TypeString,
+				Computed: true,
+			},
+			"sse_kms_key_id": &schema.Schema{
+				Type:     schema.TypeString,
+				Computed: true,
+			},
+			"storage_class": &schema.Schema{
+				Type:     schema.TypeString,
+				Computed: true,
+			},
+			"version_id": &schema.Schema{
+				Type:     schema.TypeString,
+				Optional: true,
+				Computed: true,
+			},
+			"website_redirect_location": &schema.Schema{
+				Type:     schema.TypeString,
+				Computed: true,
+			},
+		},
+	}
+}
+
+func dataSourceAwsS3ObjectRead(d *schema.ResourceData, meta interface{}) error {
+	conn := meta.(*AWSClient).s3conn
+
+	bucket := d.Get("bucket").(string)
+	key := d.Get("key").(string)
+
+	input := s3.HeadObjectInput{
+		Bucket: aws.String(bucket),
+		Key:    aws.String(key),
+	}
+	if v, ok := d.GetOk("range"); ok {
+		input.Range = aws.String(v.(string))
+	}
+	if v, ok := d.GetOk("version_id"); ok {
+		input.VersionId = aws.String(v.(string))
+	}
+
+	versionText := ""
+	uniqueId := bucket + "/" + key
+	if v, ok := d.GetOk("version_id"); ok {
+		versionText = fmt.Sprintf(" of version %q", v.(string))
+		uniqueId += "@" + v.(string)
+	}
+
+	log.Printf("[DEBUG] Reading S3 object: %s", input)
+	out, err := conn.HeadObject(&input)
+	if err != nil {
+		return fmt.Errorf("Failed getting S3 object: %s", err)
+	}
+	if out.DeleteMarker != nil && *out.DeleteMarker == true {
+		return fmt.Errorf("Requested S3 object %q%s has been deleted",
+			bucket+key, versionText)
+	}
+
+	log.Printf("[DEBUG] Received S3 object: %s", out)
+
+	d.SetId(uniqueId)
+
+	d.Set("cache_control", out.CacheControl)
+	d.Set("content_disposition", out.ContentDisposition)
+	d.Set("content_encoding", out.ContentEncoding)
+	d.Set("content_language", out.ContentLanguage)
+	d.Set("content_length", out.ContentLength)
+	d.Set("content_type", out.ContentType)
+	// See https://forums.aws.amazon.com/thread.jspa?threadID=44003
+	d.Set("etag", strings.Trim(*out.ETag, `"`))
+	d.Set("expiration", out.Expiration)
+	d.Set("expires", out.Expires)
+	d.Set("last_modified", out.LastModified.Format(time.RFC1123))
+	d.Set("metadata", pointersMapToStringList(out.Metadata))
+	d.Set("server_side_encryption", out.ServerSideEncryption)
+	d.Set("sse_kms_key_id", out.SSEKMSKeyId)
+	d.Set("storage_class", out.StorageClass)
+	d.Set("version_id", out.VersionId)
+	d.Set("website_redirect_location", out.WebsiteRedirectLocation)
+
+	if isContentTypeAllowed(out.ContentType) {
+		input := s3.GetObjectInput{
+			Bucket: aws.String(bucket),
+			Key:    aws.String(key),
+		}
+		if v, ok := d.GetOk("range"); ok {
+			input.Range = aws.String(v.(string))
+		}
+		if v, ok := d.GetOk("version_id"); ok {
+			input.VersionId = aws.String(v.(string))
```

Comment:
Should we set `input.VersionId` to be `out.VersionId` unconditionally here? That way in the case where the `version_id` argument was *not* provided by the caller we should still be sure to get the body at the same version as we got the head.

Admittedly this race is relatively minor compared to other races in Terraform -- and certainly there are lots of them -- but doesn't seem like it'd add complexity to handle it given that we already know the version id by the time we get down here.